File: Python/004.py
def is_palindrome(n):
    s, r = str(n), str(n)[::-1]
    return s == r

def solve(d):
    n, m, largest = 10 ** (d - 1), 10 ** d, 0
    for i in range(n, m):
        for j in range(i + 1, m):
            p = i * j
            if largest < p and is_palindrome(p):
                largest = p
    return largest

# d = 2
d = 3
print(solve(d))

File: C++/004.cpp
#include <iostream>
#include <cmath>
#include <string>
using namespace std;

bool is_palindrome(int n)
{
    string s = to_string(n);
    string r = string(s.rbegin(), s.rend());
    return s == r;
}

int solve(int d)
{
    int n = pow(10, d - 1);
    int m = pow(10, d);
    int largest = 0;
    for (int i = n; i < m; i++)
        for (int j = i + 1; j < m; j++) {
            int p = i * j;
            if (largest < p && is_palindrome(p))
                largest = p;
        }
    return largest;
}

int main()
{
    // int d = 2;
    int d = 3;
    cout << solve(d) << endl;
}
